src/contexts/addresses.tsx

```diff
@@ -90,161 +159,105 @@ export const AddressesContextProvider: FC<{ overrideContextValue?: PartialDeep<A
-    async (addresses: AddressState[] = [], checkingForPendingTransactions = false) => {
-      const newAddressesState = new Map(addressesState)
+      const addressesToUpdate: Address[] = []
 
       // Refresh state for only the specified addresses, otherwise refresh all addresses of the current network
-      const addressesStateToRefresh = addresses.length > 0 ? addresses : getAddressesStateOfCurrentNetwork()
-      console.log('🌈 Refreshing addresses state: ', addressesStateToRefresh)
+      const addressesStateToRefresh = addresses.length > 0 ? addresses : addressesOfCurrentNetwork
+      console.log('🌈 Fetching addresses data from API: ', addressesStateToRefresh)
 
       // The state should always update when clicking the "refresh" button, but when checking for pending transactions
       // it should only update when at least one pending transaction has been confirmed.
       let shouldUpdate = !checkingForPendingTransactions
 
-      for (const currentAddressState of addressesStateToRefresh) {
-        if (currentAddressState) {
-          const currentNumberOfPendingTx = currentAddressState.transactions.pending
-            ? currentAddressState.transactions.pending.length
-            : 0
-
-          const addressDetails = await fetchAddressDetails(currentAddressState.hash)
-          const addressConfirmedTransactions = await fetchAddressConfirmedTransactions(currentAddressState.hash)
-
-          // Filter pending addresses and remove the ones that are now confirmed
-          const addressPendingTransactions = currentAddressState.transactions.pending
-            ? currentAddressState.transactions.pending.filter(
-                (pendingTx) => !addressConfirmedTransactions.find((confirmedTx) => confirmedTx.hash === pendingTx.txId)
-              )
-            : currentAddressState.transactions.pending
-
-          const newNumberOfPendingTx = addressPendingTransactions ? addressPendingTransactions.length : 0
-
-          if (checkingForPendingTransactions && newNumberOfPendingTx !== currentNumberOfPendingTx) {
-            shouldUpdate = true
-          }
-
-          if (addressDetails) {
-            newAddressesState.set(constructMapKey(currentAddressState.hash), {
-              ...currentAddressState,
-              details: addressDetails,
-              transactions: {
-                pending: addressPendingTransactions,
-                confirmed: addressConfirmedTransactions
-              },
-              lastUsed: addressConfirmedTransactions.length > 0 ? addressConfirmedTransactions[0].timestamp : undefined
-            })
-          }
+      for (const address of addressesStateToRefresh) {
+        await address.fetchDetails(client)
+        await address.fetchConfirmedTransactions(client)
+
+        const initialNumberOfPendingTransactions = address.transactions.pending.length
+
+        // Filter pending addresses and remove the ones that are now confirmed
+        address.updatePendingTransactions()
```

Comment:
Mhh, I don't know what to think about this. Calling a function which mutate some properties and doesn't return anything (when you need a comment to explain what a function does, it's usually that something could be more explicit :D). I'm starting to wonder if we shouldn't adopt a more immutable and functional way of coding instead of always mutate the same address object...

I'll do some research and see if I can propose something different. In the meantime, let's continue like that.

Reply:
Also here I [initially implemented it by creating new objects](https://github.com/alephium/alephium-wallet/blob/master/src/contexts/addresses.tsx#L188-L197), but I felt that the code becomes very confusing, so I replaced it with instances of the `Address` class (instead of an object) and always using the `setAddress` every time an `Address` instance is being modified. It feels so much cleaner to code this way for me. Happy to discuss to find a better approach.

Reply:
>  I'm starting to wonder if we shouldn't adopt a more immutable and functional way

Here you mean "should", right?